azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -42,8 +43,10 @@ class GoalState(object):
     # fetch the goal state.
     #
     ContainerID = "00000000-0000-0000-0000-000000000000"
+    _IncarnationForCerts = None
+    _Certs = None
 
-    def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
```

Comment:
why did we remove the full_goal_state flag? seems like now WireClient._UpdateType.GoalStateForced is ignored

Reply:
Removed the comment.
I removed the flag because everything is on demand now. We retrieve the full goal state as each property is accessed.

Reply:
what was the motivation to move to an on-demand model?

Reply:
The initial motivation was a discussion we had concerning ExtensionsConfig. We decided to make this on demand to simplify the logic.
I then extended this to the other properties (except for Certs due to issues found during testing). If I recall, you said "be my guest, but be my careful guest" so I've tried to be very careful. :)

Reply:
Forgot to mention that this is also for performance. We only retrieve the documents we absolutely need.

Reply:
I'll take a look at this with more detail once I'm done with my overview. I am concerned about how the shared instance in wire_client will behave under this model and also about certs.

Reply:
Certs we always retrieve because we never actually retrieve the certs. The logic depends on us retrieving them when we retrieve the goal state. I tried explicitly retrieving the property, but couldn't fix issues I saw during the DCR runs, so I just put the code back.

Reply:
now, the call to save_goal state references those properties, triggering a fetch, so the on-demand model doesn't hold, right?

Reply:
we had a bunch of issues in the previous code because those properties were being updated as side-effects of other operations so we move to a more explicit model (given by WireClient._UpdateType)... the on-demand model seems to bring back some of those issues.

Reply:
For save_goal_state, we can probably update that logic later. I recall from the meeting you said that eventually you want to move to on-demand, so I started the work to get us there.

For certs, I added a check so we only do this once per goal state incarnation.

Reply:
>For save_goal_state, we can probably update that logic later. 

I am following the sequence of this call (update.py line 259):

```
#
# Fetch the goal state one time; some components depend on information provided by the goal state and this
# call ensures the required info is initialized (e.g telemetry depends on the container ID.)
#
protocol = self.protocol_util.get_protocol()
protocol.update_goal_state()
```
This call is not supposed to fetch any of the nested properties of the goal state (e.g. extconfig) but it triggers a call to save_goal_state, which I think would retrieve all properties.

> recall from the meeting you said that eventually you want to move to on-demand

Sorry, that may have been a misunderstanding... I did not mean on-demand, but explicit (e.g. today provisioning will fetch the entire goal state, when all it needs is the certs).

Reply:
btw, the above is just an example, but we also have similar calls on different threads

Reply:
So one option here, without rearchitecting the entire thing - would be to modify save_goal_state to only save those components that have already been retrieved. So, if we haven't read shared_conf, then save_goal_state wouldn't retrieve it.

Would that work?

Reply:
Removed the on demand properties.